(DOCSP-27011, DOCSP-27012, DOCSP-27013): Consolidate React Native sync session pages with @realm/react-ification (#2587)
mongodb/docs-realm: source/sdk/react-native/sync-data/manage-sync-session.txt

Manage a Sync Session - React Native SDK
========================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

When you use Atlas Device Sync, the Realm REact Native SDK syncs data with Atlas
in the background using a sync session. The sync session starts whenever
you open a synced realm.

Prerequisites
-------------

Before you can manage a sync session, you must perform the following:

#. :ref:`Open a synced realm <react-native-open-a-synced-realm>`
#. If you're using Flexible Sync, :ref:`add a sync subscription
   <react-native-sync-subscribe-to-queryable-fields>`
#. Wrap components that use the ``useRealm()`` hook for a synced realm
   with the ``AppProvider``, ``UserProvider``, and ``RealmProvider`` components.
   For more information on configuring and opening a synced realm,
   refer to :ref:`Open a Synced Realm <react-native-open-a-synced-realm>`.

.. _react-native-access-sync-session:

Access Sync Session
-------------------

Access a synced realm in a component with the ``useRealm()`` hook.
Access the sync session with the synced realm's :js-sdk:`Realm.syncSession
<Realm.html#syncSession>` property.

> .. literalinclude:: /examples/generated/react-native/ts/sync-session.test.snippet.sync-session.tsx

Author: i get what you mean. was thinking to show an example of sync session in action. but can do a more abstract example here, and add this one to the pause/resume sync session section.

   :language: typescript

.. _react-native-pause-or-resume-a-sync-session:

Pause or Resume a Sync Session
------------------------------

Opening a synced realm starts a :js-sdk:`sync session <Realm.App.Sync.Session.html>`.
You can pause and resume the sync session on the realm. Pausing a sync
session only pauses that realm's sync session. If you have more than one
open realm, pause does not affect the sync sessions for other realms.

To pause synchronization, use the :js-sdk:`syncSession.pause()
<Realm.App.Sync.Session.html#~pause>` method. To resume synchronization, use the
:js-sdk:`syncSession.resume() <Realm.App.Sync.Session.html#~resume>` method.

> .. literalinclude:: /examples/generated/node/sync-changes-between-devices.snippet.sync-changes-between-devices-pause-or-resume-sync-session.js

Reviewer: The `Access Sync Session` already show's how to do this in the context of React Native.  This seems to be redundant and the above example is more usable.

Check the Network Connection
----------------------------

To check the current state of the connection to the server, call the
:js-sdk:`syncSession.connectionState()
<Realm.App.Sync.Session.html#.connectionState>` method.

Realm's *offline-first* design means that you generally don't need to
check the current network connection state. That said, the
``syncSession.connectionState()`` method is available if your app needs to
gets the current state of the connection to the server.

.. literalinclude:: /examples/generated/node/sync-changes-between-devices.snippet.sync-changes-between-devices-check-network-connection.js
   :language: javascript

.. _react-native-multiplex-sync-sessions:

Multiplex Sync Sessions
-----------------------

Enable :wikipedia:`session multiplexing <Session_multiplexing>` to consolidate
multiple sync sessions of a Realm app.
Only use session multiplexing if you see errors about reaching the file descriptor limit,
and you know you are using many sync sessions.

To enable session multiplexing, call :js-sdk:`Realm.App.Sync.enableSessionMultiplexing()
<Realm.App.Sync.html#.enableSessionMultiplexing>` with your :js-sdk:`Realm.App <Realm.App.html>`.

> .. literalinclude:: /examples/SyncChanges/enable-session-multiplexing.js

Reviewer 2: While this example is correct, we should realm reactify it to show `useRealm()` and maybe even enabling multiplexing within a `useEffect()` hook, though I'm not certain that's the best way to do it.

Author: > While this example is correct, we should realm reactify it to show `useRealm()` and maybe even enabling multiplexing within a `useEffect()` hook, though I'm not certain that's the best way to do it.

this method doesn't relate to teh `useRealm` hook since it's a static method on the Realm.App.Realm class

Author: > We should ensure that this needs to happen either before or after opening `realm`. 

@takameyer, do you mean that you **cannot** call `Realm.App.Sync.enableSessionMultiplexing()` while the realm is in the process of opening?

Reviewer: I think I had been reading too much code.  The example shown is fine.  I believe I assumed you had to call it on `app`, which is not the case.

#. If you're using Flexible Sync, :ref:`add a sync subscription
   <react-native-sync-subscribe-to-queryable-fields>`
#. Wrap components that use the ``useRealm()`` hook for a synced realm
   with the ``<AppProvider>``, ``<UserProvider>``, and ``<RealmProvider>`` components.
   For more information on configuring and opening a synced realm,
   refer to :ref:`Open a Synced Realm <react-native-open-a-synced-realm>`.

.. _react-native-access-sync-session:

Access Sync Session
-------------------

Access a synced realm in a component with the ``useRealm()`` hook.
Access the sync session with the realm's :js-sdk:`Realm.syncSession
<Realm.html#syncSession>` property.

.. literalinclude:: /examples/generated/react-native/ts/access-sync-session.test.snippet.access-sync-session.tsx
   :language: typescript

.. _react-native-pause-or-resume-a-sync-session:

Pause or Resume a Sync Session
------------------------------

Opening a synced realm starts a :js-sdk:`sync session <Realm.App.Sync.Session.html>`.
You can pause and resume the realm's sync session. If you have more than one
open realm, pause does not affect the other realms' sync sessions.

> To pause synchronization, use the :js-sdk:`syncSession.pause()
> <Realm.App.Sync.Session.html#~pause>` method. To resume synchronization, use the
> :js-sdk:`syncSession.resume() <Realm.App.Sync.Session.html#~resume>` method.

Author: ```suggestion
To pause synchronization, use the :js-sdk:`Realm.syncSession.pause()
<Realm.App.Sync.Session.html#~pause>` method. To resume synchronization, use the
:js-sdk:`Realm.syncSession.resume() <Realm.App.Sync.Session.html#~resume>` method.
```

.. literalinclude:: /examples/generated/react-native/ts/pause-unpause-sync-session.test.snippet.pause-unpause-sync-session.tsx
   :language: typescript

.. |pause_func_name| replace:: ``.pause()``
.. include:: /includes/note-details-about-pause.rst

.. _react-native-check-sync-progress:

Check Upload & Download Progress for a Sync Session
---------------------------------------------------

To check the upload and download progress for a sync session, add a progress

> notification using the :js-sdk:`syncSession.addProgressNotification() <Realm.App.Sync.Session.html#.addProgressNotification>` method.
> 
> The ``syncSession.addProgressNotification()`` method takes in the following three parameters:

Author: ```suggestion
notification using the :js-sdk:`Realm.syncSession.addProgressNotification() <Realm.App.Sync.Session.html#.addProgressNotification>` method.

The ``Realm.syncSession.addProgressNotification()`` method takes in the following three parameters:
```

- A ``direction`` parameter.
  Set to ``"upload"`` to register notifications for uploading data.
  Set to ``"download"`` to register notifications for downloading data.
- A ``mode`` parameter. Set to ``"reportIndefinitely"``
  for the notifications to continue until the callback is unregistered using

>   :js-sdk:`syncSession.removeProgressNotification() <Realm.App.Sync.Session.html#.removeProgressNotification>`.

Author: ```suggestion
  :js-sdk:`Realm.syncSession.removeProgressNotification() <Realm.App.Sync.Session.html#.removeProgressNotification>`.
```

.. _react-native-manage-sync-session:

========================================
Manage a Sync Session - React Native SDK
========================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

When you use Atlas Device Sync, the React Native SDK syncs data with Atlas
in the background using a sync session. A sync session starts whenever
you open a synced realm.

Prerequisites
-------------

Before you can manage a sync session, you must perform the following:

#. :ref:`Open a synced realm <react-native-open-a-synced-realm>`
#. If you're using Flexible Sync, :ref:`add a sync subscription
   <react-native-sync-subscribe-to-queryable-fields>`
#. Wrap components that use the ``useRealm()`` hook for a synced realm
   with the ``<AppProvider>``, ``<UserProvider>``, and ``<RealmProvider>`` components.
   For more information on configuring and opening a synced realm,
   refer to :ref:`Open a Synced Realm <react-native-open-a-synced-realm>`.

.. _react-native-access-sync-session:

Access Sync Session
-------------------

Access a synced realm in a component with the ``useRealm()`` hook.

> Access the sync session with the synced realm's :js-sdk:`Realm.syncSession

Reviewer 2: I know the 'synced' here provides more specificity, but there's a lot of 'sync' in these two sentences and we already set the stage of a synced realm in the first sentence.

```suggestion
Access the sync session with the realm's :js-sdk:`Realm.syncSession
```

<Realm.html#syncSession>` property.

.. literalinclude:: /examples/generated/react-native/ts/access-sync-session.test.snippet.access-sync-session.tsx
   :language: typescript

.. _react-native-pause-or-resume-a-sync-session:

Pause or Resume a Sync Session
------------------------------

Opening a synced realm starts a :js-sdk:`sync session <Realm.App.Sync.Session.html>`.

> You can pause and resume the sync session on the realm. Pausing a sync
> session only pauses that realm's sync session. If you have more than one
> open realm, pause does not affect the sync sessions for other realms.

Reviewer 2: I think we can remove this sentence, as the following sentence gets the point across.
```suggestion
You can pause and resume the realm's sync session. If you have more than one
open realm, pause does not affect the other realms' sync sessions.
```

  ``transferred`` is the current number of bytes already transferred.
  ``transferable`` is the total number of bytes already transferred
  plus the number of bytes pending transfer.

.. include:: /includes/flex-sync-unsupported-progress-notifications.rst

The following example registers a callback on the ``syncSession`` to
listen for upload events indefinitely. The example writes to the realm and
then unregisters the ``syncSession`` notification callback.

.. tabs-realm-languages::

   .. tab::
      :tabid: typescript

      .. literalinclude:: /examples/generated/react-native/ts/check-upload-download-progress.test.snippet.check-upload-download-progress.tsx
         :language: typescript

   .. tab::
      :tabid: javascript

      .. literalinclude:: /examples/generated/react-native/js/check-upload-download-progress.test.snippet.check-upload-download-progress.jsx
         :language: javascript

.. _react-native-check-network-connection:

Check the Network Connection
----------------------------

Realm's *offline-first* design means that you generally don't need to
check the current network connection state since data syncs in the background

> when connection is available.

Reviewer 2: ```suggestion
when a connection is available.
```

That said, the Realm SDK provides methods to get the current state of the network
connection to the server.

To check the current state of the connection to the server, call
:js-sdk:`Realm.syncSession.isConnected() <Realm.App.Sync.Session.html#isConnected>`.

> This method returns a boolean that is ``true`` if there is network connection and

Reviewer 2: ```suggestion
This method returns a boolean that is ``true`` if there is a network connection and
```

the sync session is active.

> To listen for connection state changes, call :js-sdk:`Realm.syncSession.addConnectionNotification()

Reviewer 2: The link text for these gets quite long and kind of hard to read. Could we shorten to just the method and let users follow the link for further details?

We're inconsistent throughout the page. In the previous section, we omit the `Realm.` part. Maybe `syncSession.<method>` is a good compromise?

Author: will shift these to be just `SyncSession.<method>()`
